### PR TITLE
Feature/tags

### LINKED ISSUE #25: Tags
- [x] on_message
- [x] tag create
- [x] tag delete
- [x] tag toggle
- [x] tag find
- [x] tag list
- [x] reply list
- [x] reply toggle

### PR DESCRIPTION
**Pull request by NathanealV**
- [Added tag create, delete, list, and toggle commands](https://github.com/ib-ai/ib.py/commit/821d44af16e6ade536867d0f3e0fce09878f16d7)
- [Added tags to cached.py](https://github.com/ib-ai/ib.py/commit/588fedc749fc58a475518a60d6f8067d9eb6f9e9)
- [Changed RegexConverter to class for better typing](https://github.com/ib-ai/ib.py/commit/7215b79409e8da2399078bb1658a6224bb2ce9da)
- [Added disableReply and GuildReply tables](https://github.com/ib-ai/ib.py/commit/e720efa0ba9f651ca31d2fa6e146bb3755ec4035)

### CODE REVIEW

File: db/models.py [hunk 1]
```diff
@@ -87,6 +87,13 @@ class Meta():
     finished = fields.BooleanField(default=False)
     vote_ladder_id = fields.OneToOneField('models.GuildVoteLadder')
 
+class GuildReply(Model):
```
Comment (ilikecubesnstuff) [inline db/models.py]: Something like `GuildSuppressReply` is a more intuitive name.

File: utils/converters.py
Comment (ilikecubesnstuff) [inline utils/converters.py:1]: It could be worth changing all the converters into classes for consistency's sake.

File: db/models.py [hunk 2]
```diff
@@ -87,6 +87,13 @@ class Meta():
...
+    class Meta():
+        table = "reply"
+
+    reply_id = fields.BigIntField(pk=True, unique=True)
+    channels = ArrayField(element_type="bigint", null = True)
```
Comment (ilikecubesnstuff) [inline db/models.py]: Keyword argument spacing.

File: db/cached.py [hunk 3]
```diff
@@ -1,18 +1,28 @@
...
 
+
 @alru_cache
 async def get_all_filters() -> list[StaffFilter]:
     return await StaffFilter.all()
+
+
+@alru_cache
+async def get_all_tags() -> list[StaffTag]:
+    return await StaffTag.all()
```
Comment (ilikecubesnstuff) [inline db/cached.py]: These share a pattern - the code could be condensed with a factory function, e.g.
```py
def model_cache_factory(Model):
    @alru_cache
    async def model_cache() -> list[Model]:
        return await Model.all()
    return model_cache
```

File: db/models.py (hunk 1 above)
Comment (ilikecubesnstuff) [reply db/models.py]: Actually, I think this can be combined with the already existing `GuildData` model.

File: cogs/tags.py [hunk 4]
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
...
-    async def create(self, ctx: commands.Context):
+    async def create(self, ctx: commands.Context, trigger: RegexConverter, output: str):
         """
         Create a tag.
         """
-        raise NotImplementedError('Command requires implementation and permission set-up.')
-    
+        if len(trigger) > NAME_SIZE_LIMIT - 10: # 2 for backticks and 8 for [Disabled]
+            await ctx.send(f"The tag trigger is too long (should not exceed {NAME_SIZE_LIMIT - 2} characters).")
+            return
```
Comment (ilikecubesnstuff) [inline cogs/tags.py]: Inconsistency between the if statement and reply?

File: cogs/tags.py [hunk 5]
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
...
+            await ctx.send(f"The tag output is too long (should not exceed {VALUE_SIZE_LIMIT} characters).")
+            return
+
+        tag = await StaffTag.filter(trigger=trigger).get_or_none()
+
+        if tag:
+            tag.output = output
+            await tag.save()
+        else:
+            await StaffTag.create(trigger=trigger, output=output)
```
Comment (ilikecubesnstuff) [inline cogs/tags.py]: Compress this with `update_or_create`.

File: cogs/tags.py [hunk 6]
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
...
+        List of tags with specified filter, or all tags if none specified.
         """
-        List of all tags.
-        """ 
-        raise NotImplementedError('Command requires implementation and permission set-up.')
-    
+        tags = await get_all_tags()
+
+        if trigger:
+            trigger = (await RegexConverter().convert(ctx, trigger)).lower()
```
Comment (ilikecubesnstuff) [inline cogs/tags.py]: Use `RegexConverter` in the function signature as a type annotation.

File: cogs/tags.py [hunk 7]
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
...
+            tags = [tag for tag in tags if trigger and trigger in tag.trigger.lower()]
```
Comment (ilikecubesnstuff) [inline cogs/tags.py]: The truthiness of `trigger` is already checked, `trigger in tag.trigger.lower()` should be sufficient.

File: cogs/tags.py [hunk 8]
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
...
+
+        # If bot, return
+        if message.author.bot:
+            return
+
+        # If bot replies are disabled for channel, return
+        disable_reply = message.channel.id in (await GuildReply.get_or_create(reply_id = message.guild.id))[0].channels
+
+        if disable_reply:
+            return
```
Comment (ilikecubesnstuff) [inline cogs/tags.py]: If you merge `GuildReply` into `GuildData`, this part could be cleaner:
```py
guild_data = await get_guild_data(guild_id = message.guild.id)
if message.channel.id in guild_data.suppressed_replies:
    return
```

File: cogs/tags.py [hunk 9]
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
...
+
+        tags = await get_all_tags()
+
+        for tag in tags:
+            if tag.disabled:
+                continue
+
+            if re.fullmatch(tag.trigger.lower(), message.content.lower(), re.IGNORECASE):
```
Comment (ilikecubesnstuff) [inline cogs/tags.py]: I suggest keeping the tag's trigger in its original case. Wait for another opinion, but I remember the lowercasing of triggers causing unexpected behaviors for some regular expressions with special characters (like `\W`, for example).

File: cogs/tags.py [hunk 10]
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
...
+
+        if not tag:
+            await ctx.send(f"Tag `{trigger}` does not exist.")
+            return
+
+        await tag.delete()
+        get_all_tags.cache_clear()
+
+        logger.debug(f"Deleted tag {trigger}.")
+        await ctx.send(f"Deleted tag `{trigger}`.")
```
Comment (ilikecubesnstuff) [inline cogs/tags.py]: Adding the tag's output in the log and reply would be handy.

File: cogs/tags.py [hunk 11]
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
...
-        raise NotImplementedError('Command requires implementation and permission set-up.')
-    
+        tags = await get_all_tags()
+
+        if trigger:
+            trigger = (await RegexConverter().convert(ctx, trigger)).lower()
+            tags = [tag for tag in tags if trigger and trigger in tag.trigger.lower()]
+
+        names = [
+            f"{'[Disabled]' if tag.disabled else ''} `{tag.trigger}`" for tag in tags]
```
Comment (ilikecubesnstuff) [inline cogs/tags.py]: This appears to leave a space at the start of non-disabled tags. Is an f-string like ```f"{'[Disabled] ' if tag.disabled else ''}`{tag.trigger}`"``` what is intended?

File: cogs/tags.py [hunk 12]
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
...
+        if disabled:
+            guild_reply.channels.remove(channel.id)
+        else:
+            guild_reply.channels.append(channel.id)
+
+        await guild_reply.save()
+
+        logger.debug(
+            f"Channel {channel.id} replies are now {'disabled' if not disabled else 'enabled'}.")
+        await ctx.send(f"Replies for <#{channel.id}> are now {'disabled' if not disabled else 'enabled'}.")
```
Comment (ilikecubesnstuff) [inline cogs/tags.py]: Perhaps move these into `guilddata.py` if you merge `GuildReply` with `GuildData`.

File: cogs/tags.py [hunk 13]
```diff
@@ -11,51 +24,159 @@ async def on_message(self, message: discord.Message):
...
+
+    @reply.command()
+    async def toggle(self, ctx: commands.Context, channel: discord.TextChannel):
+        """
+        Toggle if bot replies are disabled for specified channel.
+        """
+        guild_reply = (await GuildReply.get_or_create(reply_id = ctx.guild.id))[0]
+
+        if not guild_reply.channels:
+            guild_reply.channels = []
```
Comment (ilikecubesnstuff) [inline cogs/tags.py]: Redundant?

**Review by ilikecubesnstuff — CHANGES_REQUESTED** [review]
Thank you for implementing the tags feature! My main feedback is about merging the new `GuildReply` model with the existing `GuilData` model. Apart from that, I have a bunch more small comments in various places.

File: cogs/tags.py (hunk 12 above)
Comment (NathanealV) [reply cogs/tags.py]: Can be moved when the moderation PR have been merged - ignoring for now.

**Comment by NathanealV** [comment]
- Removed `.lower()` when checking for tag triggers
- Consistency with checking for tag trigger length
- Replaced redundant code with `update_or_create`
- Adjusted logging and success message for tag delete
- Adjusted `reply` commands for changes to DB schema
- Formatting

File: cogs/tags.py [hunk 14]
```diff
@@ -55,34 +56,36 @@ async def tag(self, ctx: commands.Context):
...
+    async def tag_create(self, ctx: commands.Context, trigger: RegexConverter,
+                         output: str):
         """
         Create a tag.
         """
-        if len(trigger) > NAME_SIZE_LIMIT - 10: # 2 for backticks and 8 for [Disabled]
-            await ctx.send(f"The tag trigger is too long (should not exceed {NAME_SIZE_LIMIT - 2} characters).")
+        if len(
+                trigger
+        ) > NAME_SIZE_LIMIT - 11:  # 2 for backticks and 9 for [Disabled] with trailing whitespace
```
Comment (ilikecubesnstuff) [inline cogs/tags.py]: Doesn't `[Disabled] ` take up 11 characters? So including the 2 backticks, this should total 13?

**Review by ilikecubesnstuff — APPROVED** [review]

**Pull request merged**